cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -1,12 +1,80 @@
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
+
 use async_trait::async_trait;
-use catalog_api_v1::types::{self as api_types, error as api_error};
-use catalog_api_v1::{Client as APIClient, Error as APIError};
+use catalog_api_v1::types::{self as api_types, error as api_error, ErrorResponse};
+use catalog_api_v1::{Client as APIClient, Error as APIError, ResponseValue};
 use enum_dispatch::enum_dispatch;
+use reqwest::header::HeaderMap;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
 use crate::data::System;
+use crate::models::search::SearchResults;
 
 pub const DEFAULT_CATALOG_URL: &str = "https://flox-catalog.flox.dev";
+pub const FLOX_CATALOG_MOCK_DATA_VAR: &str = "_FLOX_USE_CATALOG_MOCK";
+
+type ResolvedGroups = Vec<ResolvedPackageGroup>;
+
+// Arc allows you to push things into the client from outside the client if necessary
+// Mutex allows you to share across threads (necessary because of tokio)
+// RefCell lets us mutate the field without needing to make the Client trait methods mutable
+type MockField<T> = Arc<Mutex<RefCell<T>>>;
+
+/// A generic response that can be turned into a [ResponseValue]. This is only necessary for
+/// representing error responses.
+// TODO: we can handle headers later if we need to
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GenericResponse<T> {
+    pub(crate) inner: T,
+    pub(crate) status: u16,
+}
```

Comment:
question: Do we need this because the progenitor/reqwest client will map error responses to a non serializable Error type?

Reply:
I'm using this specifically because the `ResponseValue<T>` doesn't implement the serialization traits. In order to fill out an `APIError::ErrorResponse` you need a `ResponseValue<T>`, which includes headers, status code, etc. The `ErrorResponse` type is serializable, but doesn't contain much context. Being able to serialize a generic response value also just seems useful.